src/smt.rs

```diff
@@ -1,163 +1,632 @@
     pub name: String,
+    pub sort: SMTSort,
 }
 
-impl SMTFormat for SMTVariable {
-    fn as_smt(&self) -> String {
-        self.name.clone()
+impl SMTVariable {
+    pub fn new(name: String, sort: SMTSort) -> Self {
+        SMTVariable { name, sort }
     }
 }
 
-pub trait SMTFormat {
-    fn as_smt(&self) -> String;
+/// We keep the SMT expressions loose and runtime based for now.
+/// A potential future refactor may add compile time guarantees
+/// that illegal expressions cannot be built.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SMTExpr {
+    pub op: SMTOp,
+    pub args: Vec<SMTExpr>,
 }
 
-type BoxedSMTFormat = Box<dyn SMTFormat>;
+impl From<u64> for SMTExpr {
+    fn from(input: u64) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Literal(format!("{:064x}", input)),
+            args: vec![],
+        }
+    }
+}
 
-impl SMTFormat for String {
-    fn as_smt(&self) -> String {
-        self.to_string()
+impl From<SMTVariable> for SMTExpr {
+    fn from(input: SMTVariable) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Variable(input),
+            args: vec![],
+        }
     }
 }
 
-impl SMTFormat for usize {
-    fn as_smt(&self) -> String {
-        format!("#x{:064x}", self)
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTStatement {
+    Assert(SMTExpr),
+    DeclareConst(SMTVariable),
+    DeclareFun(SMTVariable, Vec<SMTSort>),
+    DefineConst(SMTVariable, SMTExpr),
+    DefineFun(SMTVariable, Vec<SMTSort>, SMTExpr),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTOp {
+    Eq,
+    Not,
+    And,
+    Or,
+    Ite,
+    BvNot,
+    BvAnd,
+    BvOr,
+    BvXor,
+    BvAdd,
+    BvSub,
+    BvMul,
+    BvUDiv,
+    BvSDiv,
+    BvURem,
+    BvSMod,
+    BvULt,
+    BvUGt,
+    BvSLt,
+    BvSGt,
+    BvShl,
+    BvLShr,
+    BvAShr,
+    Extract(u64, u64),
+    Concat,
+    Select,
+    Store,
+    Literal(String),
+    Variable(SMTVariable),
+    UF(SMTVariable), // TODO We should have a specialized SMTFunction
+}
+
+// This is a hack to get #x000000000000000000000000 needed to clear the
+// first 12 bytes of an address.
+// Maybe we should just declare address as 20 bytes in the first place
+// and upcast when needed.
+
+pub struct Literal12Bytes(u64);
+
+impl From<u64> for Literal12Bytes {
+    fn from(input: u64) -> Self {
+        Literal12Bytes(input)
     }
 }
 
-impl<T: SMTFormat> SMTFormat for Vec<T> {
-    fn as_smt(&self) -> String {
-        self.iter()
-            .map(SMTFormat::as_smt)
-            .collect::<Vec<_>>()
-            .join(" ")
+impl From<Literal12Bytes> for SMTExpr {
+    fn from(input: Literal12Bytes) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Literal(format!("{:024x}", input.0)),
+            args: vec![],
+        }
     }
 }
 
-pub enum SMTIdentifier {
-    Simple(String),
+// SMT expression builders
+
+pub fn eq(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Eq,
+        args: vec![lhs, rhs],
+    }
 }
```

Comment:
Have you thought about the following? This would remove all the `.into()` calls I think.
```suggestion
pub fn eq(lhs: Into<SMTExpr>, rhs: Into<SMTExpr>) -> SMTExpr {
    SMTExpr {
        op: SMTOp::Eq,
        args: vec![lhs.into(), rhs.into()],
    }
}
```

Reply:
Actually the u64 intos are not needed. Are you talking about the SMTVariable intos?